### PR TITLE
Created a new branch "image analysis" for image overlapping functionalities

### PR DESCRIPTION
**Pull request by MujtabaNasir**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -17,7 +17,7 @@ The project aims to provide a python package using poetry which has 3 sub-packag
 ## Installation
 1. Install mujtaba-charm package using pip:
 ```bash
-pip install git+https://github.com/MujtabaNasir/mujtaba-charm.git
+pip install git+
```
Comment (mdsrqbl) [inline README.md]: ?

File: dev-requirements.txt [hunk 2]
```diff
@@ -5,3 +5,5 @@ coverage==7.6.0
 pytest-cov==5.0.0
 sphinx==7.1.2
 sphinx-autodoc-typehints==2.0.0
+numpy==1.24.4
+matplotlib==3.9.0
```
Comment (mdsrqbl) [inline dev-requirements.txt]: why must these 2 be in 'dev'?

File: mujtaba_charm/analysis/image.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
+import matplotlib.pyplot as plt
+import numpy as np
+import tensorflow as tf
```
Comment (mdsrqbl) [inline mujtaba_charm/analysis/image.py]: this would crash the import of mujtaba_charm library in a new environment

File: mujtaba_charm/mnist.npz
Comment (mdsrqbl) [inline mujtaba_charm/mnist.npz:1]: why must you push non-code files into your python library?

File: mujtaba_charm/fonts/Ubuntu-Regular.ttf
Comment (mdsrqbl) [inline mujtaba_charm/fonts/Ubuntu-Regular.ttf:1]: why must you push non-code files into your python library?
this is an asset required by just one function, either add a runtime downloader function or leave the font selection to the user. Do not push data files to your repository.

File: mujtaba_charm/fonts/Roboto-Regular.ttf
Comment (mdsrqbl) [inline mujtaba_charm/fonts/Roboto-Regular.ttf:1]: why must you push non-code files into your python library?
this is an asset required by just one function, either add a runtime downloader function or leave the font selection to the user. Do not push data files to your repository.

File: mujtaba_charm/fonts/OpenSans-Regular.ttf
Comment (mdsrqbl) [inline mujtaba_charm/fonts/OpenSans-Regular.ttf:1]: why must you push non-code files into your python library?
this is an asset required by just one function, either add a runtime downloader function or leave the font selection to the user. Do not push data files to your repository.

File: mujtaba_charm/analysis/image.py (hunk 3 above)
Comment (mdsrqbl) [reply mujtaba_charm/analysis/image.py]: also tensorflow sucks
use pytorch

File: mujtaba_charm/analysis/image.py [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+
+
+def fetch_mnist() -> tuple[np.ndarray, np.ndarray]:
+    """
+    Fetches the MNIST dataset and saves it to a compressed file.
+
+    Returns:
+        tuple: A tuple containing the training images and labels.
+    """
+    mnist = tf.keras.datasets.mnist.load_data()
```
Comment (mdsrqbl) [inline mujtaba_charm/analysis/image.py]: use pytorch instead
(add it to your core dependencies)

File: mujtaba_charm/analysis/image.py [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+    return train_images, train_labels
+
+
+def create_digit_image(digit: int, font_path: str, image_size: tuple[int, int] = (28, 28)) -> Image.Image:
+    """
+    Creates an image of a digit using a specified font.
+
+    Args:
+        digit (int): The digit to draw.
+        font_path (str): Path to the font file.
```
Comment (mdsrqbl) [inline mujtaba_charm/analysis/image.py]: can you use font name instead of/along with path? (add a few supported fonts, use 1 of them as default)

File: mujtaba_charm/analysis/image.py [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+        image_size (tuple[int, int], optional): Size of the output image. Defaults to (28, 28).
+
+    Returns:
+        Image.Image: An image of the digit.
+    """
+    font = ImageFont.truetype(font_path, 24)
```
Comment (mdsrqbl) [inline mujtaba_charm/analysis/image.py]: let the user decide font size (use a default value too)

File: mujtaba_charm/analysis/image.py [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+    image = Image.new("L", image_size, color=255)
```
Comment (mdsrqbl) [inline mujtaba_charm/analysis/image.py]: let the user decide background/foreground color

File: mujtaba_charm/analysis/image.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+    draw = ImageDraw.Draw(image)
+    text_bbox = draw.textbbox((0, 0), str(digit), font=font)
+    text_width, text_height = text_bbox[2] - text_bbox[0], text_bbox[3] - text_bbox[1]
+    text_x = (image_size[0] - text_width) / 2
+    text_y = (image_size[1] - text_height) / 2
```
Comment (mdsrqbl) [inline mujtaba_charm/analysis/image.py]: add a comment on what you are calculating

File: mujtaba_charm/analysis/image.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+    Finds the overlapping areas between two images.
+
+    Args:
+        img1 (np.ndarray): The first image.
+        img2 (np.ndarray): The second image.
+
+    Returns:
+        np.ndarray: A binary image showing the overlapping areas.
+    """
+    mnist_array = np.array(img1)
```
Comment (mdsrqbl) [inline mujtaba_charm/analysis/image.py]: why is this variable called mnist?

File: mujtaba_charm/analysis/image.py [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+    font_array = np.array(img2)
```
Comment (mdsrqbl) [inline mujtaba_charm/analysis/image.py]: how does this function decide that img2 is a font image??

File: mujtaba_charm/analysis/image.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    text_width, text_height = text_bbox[2] - text_bbox[0], text_bbox[3] - text_bbox[1]
+    text_x = (image_size[0] - text_width) / 2
+    text_y = (image_size[1] - text_height) / 2
+    draw.text((text_x, text_y), str(digit), font=font, fill=0)
+    return image
+
+
+def get_image_overlapping(img1: np.ndarray, img2: np.ndarray) -> np.ndarray:
+    """
+    Finds the overlapping areas between two images.
```
Comment (mdsrqbl) [inline mujtaba_charm/analysis/image.py]: let the developer know what values it considers to be overlapping

File: notebooks/mnist.npz
Comment (mdsrqbl) [inline notebooks/mnist.npz:1]: why did you push this file to the repository? do you want to maintain a version history of this .npz file??

File: pyproject.toml [hunk 12]
```diff
@@ -7,9 +7,13 @@ readme = "README.md"
...
 [tool.poetry.dependencies]
 python = ">=3.8,<3.13"
+numpy = [
+   {version = "<=1.24.*", python = "<3.9"},
+   {version = ">=1.25.*", python = ">=3.9"}
+]
 
 [tool.poetry.extras]
-dev = ["pytest", "black", "isort", "coverage", "pytest-cov", "sphinx", "sphinx-autodoc-typehints"]
+dev = ["pytest", "black", "isort", "coverage", "pytest-cov", "sphinx", "sphinx-autodoc-typehints","matplotlib"]
```
Comment (mdsrqbl) [inline pyproject.toml:17]: one of your core functions is using matplotlib
so why is it in the dev dependencies?

File: pyproject.toml [hunk 13]
```diff
@@ -19,13 +23,15 @@ black = "^24.4.2"
 isort = "^5.13.2"
 coverage = "^7.6.0"
 pytest-cov = "^5.0.0"
+wheel = "^0.43.0"
```
Comment (mdsrqbl) [inline pyproject.toml:27]: why are _you_ specifying the wheel package version?

File: mujtaba_charm/analysis/image.py [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+        image_size (tuple[int, int], optional): Size of the output image. Defaults to (28, 28).
+
+    Returns:
+        Image.Image: An image of the digit.
+    """
+    font = ImageFont.truetype(font_path, 24)
+    image = Image.new("L", image_size, color=255)
+    draw = ImageDraw.Draw(image)
+    text_bbox = draw.textbbox((0, 0), str(digit), font=font)
+    text_width, text_height = text_bbox[2] - text_bbox[0], text_bbox[3] - text_bbox[1]
```
Comment (mdsrqbl) [inline mujtaba_charm/analysis/image.py]: add `breakpoint()` here and use pdb to inspect the values of these variables

**Review by mdsrqbl — CHANGES_REQUESTED** [review]

**Comment by mdsrqbl** [comment]
your PR name should describe the functionality or the changes made, not the git operations you did.
